storage/range.go

```diff
@@ -1759,27 +1760,64 @@ func (r *Range) AdminMerge(args *proto.AdminMergeRequest, reply *proto.AdminMerg
 	}
 }
 
-// InternalChangeReplicas is called after a raft configuration change has committed.
-func (r *Range) InternalChangeReplicas(batch engine.Engine,
-	req *proto.InternalChangeReplicasRequest,
-	reply *proto.InternalChangeReplicasResponse) {
-	r.Lock()
-	defer r.Unlock()
-	// Apply the committed membership change to r.Desc.
-	// TODO(bdarnell): update the range addressing records too (atomically).
-	// Need to either combine this with EndTransaction or queue up a transaction on
-	// the meta records.
-	r.Desc.Replicas = r.Desc.Replicas[:0]
-	for _, n := range req.Nodes {
-		nodeID, storeID := decodeRaftNodeID(multiraft.NodeID(n))
-		r.Desc.Replicas = append(r.Desc.Replicas, proto.Replica{
-			NodeID:  nodeID,
-			StoreID: storeID,
-			// TODO(bdarnell): Set attributes. Copy from gossip or pass through request?
-		})
+// ChangeReplicas adds or removes a replica of a range. The change is performed
+// in a distributed transaction and takes effect when that transaction is committed.
+// When removing a replica, only the NodeID and StoreID fields of the Replica are used.
+func (r *Range) ChangeReplicas(changeType proto.ReplicaChangeType, replica proto.Replica) error {
+	// Only allow a single change per range at a time.
+	if !atomic.CompareAndSwapInt32(&r.metaLock, int32(0), int32(1)) {
+		return util.Errorf("range %d metadata locked", r.Desc.RaftID)
+	}
+	defer func() { atomic.StoreInt32(&r.metaLock, int32(0)) }()
+
+	// Validate the request and prepare the new descriptor.
+	updatedDesc := *r.Desc
+	found := -1
+	for i, existingRep := range r.Desc.Replicas {
+		if existingRep.NodeID == replica.NodeID && existingRep.StoreID == replica.StoreID {
+			found = i
+			break
+		}
+	}
+	if changeType == proto.ADD_REPLICA {
```

Comment:
I think this algorithm will work for adding and removing. But keep in mind that we have an issue here because we need to be precise about which replica is being replaced. Our zone configs specify attributes for replica indexes. We can't just let the replicas get jumbled up. If we always add a new replica in order to replace an existing, this feels like it will work if the order is always Add, then Remove, with no chance for another replica to be added first.